syntax = "proto3";

import "google/protobuf/timestamp.proto";


// Version maps klio config version to message version when actually
// handling/parsing klio messages
enum Version {
    UNKNOWN = 0;  // default (per proto3 requirements)
    V1 = 1;  // for entity_id+payload
    V2 = 2;  // for element+payload
}


message KlioMessage {
    message Metadata {
        // optional - used for bottom-up
        repeated KlioJob downstream = 1;
        // optional - unique jobs this message has visited
        repeated KlioJob visited = 2;
        // optional - audit log for all jobs message has visited (complete)
        repeated KlioJobAuditLogItem job_audit_log = 3;
        // optional - force computation even if output exists; default = false
        bool force = 4;
        // optional - if true, then no transformation work is done; default = false
        bool ping = 5;
    }

    message Data {
        // required for v1 messages
        string entity_id = 1;
        // optional - (v1) used for non-klio messages, (v2) used for intra-pipeline state
        bytes payload = 2;
        // required for v2 messages
        bytes element = 3;
    }

    Metadata metadata = 1;
    Data data = 2;
    // message version
    Version version = 3;
}


message KlioJob {
    message JobInput {
        string topic = 1;
        string subscription = 2;
        string data_location = 3;
    }
    // required
    string job_name = 1;
    // required
    string gcp_project = 2;
    // required
    repeated JobInput inputs = 3;
}


message KlioJobAuditLogItem {
    // required - timestamp of when the log item was created
    google.protobuf.Timestamp timestamp = 1;
    // required
    KlioJob klio_job = 2;
}
